{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cairo\n",
    "import re\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#input fasta file of genes\n",
    "#input text file with motifs, one per line\n",
    "\n",
    "genes_txt = \"INSR.fasta\"\n",
    "motifs_txt = \"motifs.txt\"\n",
    "output_folder = \"Images/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#initailize motif dictionary \n",
    "#key = given motif \n",
    "#value = regex search \n",
    "\n",
    "#IUPAC codes https://www.bioinformatics.org/sms/iupac.html\n",
    "\n",
    "nuc_codes = {}\n",
    "nuc_codes[\"A\"] = \"[Aa]\"\n",
    "nuc_codes[\"T\"] = \"[Tt]\"\n",
    "nuc_codes[\"G\"] = \"[Gg]\"\n",
    "nuc_codes[\"C\"] = \"[Cc]\"\n",
    "nuc_codes[\"R\"] = \"[AaGg]\"\n",
    "nuc_codes[\"Y\"] = \"[CcTt]\"\n",
    "nuc_codes[\"S\"] = \"[GgCc]\"\n",
    "nuc_codes[\"W\"] = \"[AaTt]\"\n",
    "nuc_codes[\"K\"] = \"[GgTt]\"\n",
    "nuc_codes[\"M\"] = \"[AaCc]\"\n",
    "nuc_codes[\"B\"] = \"[CcGgTt]\"\n",
    "nuc_codes[\"D\"] = \"[AaGgTt]\"\n",
    "nuc_codes[\"H\"] = \"[AaCcTt]\"\n",
    "nuc_codes[\"V\"] = \"[AaCcGg]\"\n",
    "nuc_codes[\"N\"] = \"[AaCcGgTt]\"\n",
    "\n",
    "motifs = {}\n",
    "\n",
    "with open(motifs_txt, \"r\") as fh:\n",
    "    for line in fh:\n",
    "        line = line.strip()\n",
    "        motif = \"\"\n",
    "        for char in line:\n",
    "            if char.capitalize() in nuc_codes:\n",
    "                motif += nuc_codes[char.capitalize()]\n",
    "            else:\n",
    "               quit(\"One of the given motifs contains an unidentified IUPAC nucleotide code\")\n",
    "        motifs[line] = motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INSR\n",
      "abcd\n"
     ]
    }
   ],
   "source": [
    "#initialize dictionary with genes\n",
    "#key = gene names\n",
    "#value = gene sequence\n",
    "\n",
    "gene_info = {}\n",
    "gene_len = {}\n",
    "\n",
    "read = \"\"\n",
    "first = True\n",
    "with open(genes_txt, \"r\") as gene:\n",
    "    for line in gene:\n",
    "        line = line.strip()\n",
    "        if line[0] != \">\":\n",
    "            read += line\n",
    "        else:\n",
    "            if first != True:\n",
    "                gene_info[gene_name] = read\n",
    "                read = \"\"\n",
    "            else:\n",
    "                first = False\n",
    "            gene = re.search(\">(\\S+)\", line)\n",
    "            gene_name = gene.group(1)\n",
    "            print(gene_name)\n",
    "            startLoc = re.search(\".+ (.+:[0-9]+-[0-9]+)\", line)\n",
    "            gene_len[gene_name] = startLoc.group(1)\n",
    "                \n",
    "gene_info[gene_name] = read\n",
    "#print(gene_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def exon_loc(string):\n",
    "    first = True\n",
    "    counter = 0\n",
    "    exon_num = 0\n",
    "    exons = {}\n",
    "\n",
    "    for char in string:\n",
    "        counter += 1\n",
    "        if ord(char) >= 65 and ord(char) <= 90:\n",
    "            if first == True:\n",
    "                exon_num += 1\n",
    "                start = counter\n",
    "            else:\n",
    "                end = counter\n",
    "            first = False\n",
    "        else:\n",
    "            if first == False:\n",
    "                exons[exon_num] = (start, end)\n",
    "            first = True\n",
    "\n",
    "    return(exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cairo_init(name):\n",
    "    width, height = 800, 500\n",
    "\n",
    "    #create the coordinates to display your graphic, desginate output\n",
    "    surface = cairo.SVGSurface(name,width, height)\n",
    "    #create the coordinates you will be drawing on (like a transparency) - you can create a transformation matrix\n",
    "    context = cairo.Context(surface)\n",
    "    #context.scale(width,height) #will set your drawing surface to a 0.0-1.0 scale\n",
    "    context.scale(width,height)\n",
    "    \n",
    "    return surface, context"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixed = (.9,.6,0), (.8,0.1,0.1), (.6,0.2,0.8), (0.1,0.9,0.9), (.4, 0.9, 0.2), (.0, 0.0, 0.9), (.9, 0.9,0), (0.9,0.3,0.8), (.5, 0.5,0.5), (.3,0.3,0.9) \n",
    "blues = (.5,1,1), (0,.2,1), (.4,.8,.6), (0,.6,.8), (.4,.8,1), (0,1,.2), (.4,.9,.7), (0,.8,.5), (.4,.4,.8), (0,.2,1)\n",
    "reds = (1,0,0), (1,0,.5), (1,.4,1), (.8,.2,.5), (.6,.2,.9), (1,.4,0), (.6,.4,.7), (0.9,.5,.5), (.7,.4,1), (1,.4,.6)\n",
    "\n",
    "for gene in gene_info:    \n",
    "    #cairo drawing gene length\n",
    "    surface, context = cairo_init(output_folder+gene+\".svg\")\n",
    "    context.set_line_width(.005)\n",
    "    context.move_to(.05,.7)        #(x,y)\n",
    "    context.line_to(.95,.7)\n",
    "    context.stroke()\n",
    "    \n",
    "    #scaling and shift for equal image size and location\n",
    "    #multiplied by .9 to account for length of gene line\n",
    "    scale = (1/len(gene_info[gene])) * .9\n",
    "    shift = .05\n",
    "    key_count = -1\n",
    "    \n",
    "    #exon location\n",
    "    exons = exon_loc(gene_info[gene])\n",
    "    for exon in exons:\n",
    "        context.rectangle((exons[exon][0]*scale) + shift,.65,(exons[exon][1]-exons[exon][0])*scale,.1)        #(x0,y0,width,height)\n",
    "        context.fill()\n",
    "    \n",
    "    #gene title\n",
    "    context.select_font_face(\"Purisa\", cairo.FONT_SLANT_NORMAL, \n",
    "        cairo.FONT_WEIGHT_NORMAL)\n",
    "    context.set_font_size(.1)\n",
    "    \n",
    "    context.move_to(.05, .55)\n",
    "    context.show_text(gene)\n",
    "    \n",
    "    #starting key with title and exon\n",
    "    context.set_font_size(.06)\n",
    "    context.move_to(.7, .1)\n",
    "    context.show_text(\"Motifs\")\n",
    "    key_pos = .1\n",
    "    key_pos += .05\n",
    "    context.set_font_size(.03)\n",
    "    context.move_to(.7, key_pos)\n",
    "    context.show_text(\"EXON\")\n",
    "    \n",
    "    #gene position\n",
    "    context.move_to(.05, .6)\n",
    "    context.show_text(gene_len[gene])\n",
    "    \n",
    "    for key in motifs:\n",
    "        key_count += 1        \n",
    "        motif_loc_start = .675\n",
    "        motif_loc_end = .725\n",
    "        kmerStart = 0\n",
    "        kmerEnd = kmerStart + len(key)\n",
    "        context.set_line_width(scale * len(key))\n",
    "        \n",
    "        while kmerEnd <= len(gene_info[gene]):    \n",
    "            if (re.match(motifs[key], gene_info[gene][kmerStart:kmerEnd])):\n",
    "                context.move_to((kmerStart * scale) + shift, motif_loc_start)\n",
    "                context.set_source_rgb(colors[key_count][0], colors[key_count][1], colors[key_count][2])\n",
    "                context.line_to((kmerStart * scale) + shift, motif_loc_end)\n",
    "                context.stroke()\n",
    "                 \n",
    "            kmerStart += 1\n",
    "            kmerEnd = kmerStart + len(key)\n",
    "            \n",
    "        #input genes into key\n",
    "        key_pos += .05\n",
    "        context.set_source_rgb(colors[key_count][0], colors[key_count][1], colors[key_count][2])\n",
    "        context.set_font_size(.03)\n",
    "        context.move_to(.7, key_pos)\n",
    "        context.show_text(key.upper())\n",
    "    surface.finish()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n"
     ]
    }
   ],
   "source": [
    "colors = []\n",
    "count = 0\n",
    "for i in range(1,11):\n",
    "    count += 1\n",
    "    red = random.uniform(0,1)\n",
    "    green = random.uniform(0,1)\n",
    "    blue = random.uniform(0,1)\n",
    "    color = red, green, blue\n",
    "    colors.append(color)\n",
    "    \n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "width, height = 800, 500\n",
    "\n",
    "#create the coordinates to display your graphic, desginate output\n",
    "surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, width, height)\n",
    "#create the coordinates you will be drawing on (like a transparency) - you can create a transformation matrix\n",
    "context = cairo.Context(surface)\n",
    "#context.scale(width,height) #will set your drawing surface to a 0.0-1.0 scale\n",
    "context.scale(width,height)\n",
    "context.set_line_width(.005)\n",
    "context.move_to(.05,.7)        #(x,y)\n",
    "context.line_to(.95,.7)\n",
    "context.stroke()\n",
    "surface.write_to_png(\"example.jpeg\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
